### PR TITLE
[5.5] support milliseconds for date_format validation rule

### PR DESCRIPTION
**Pull request by haakym**
For issue https://github.com/laravel/framework/issues/19445.

The `v` character representing milliseconds was introduced in PHP 7 for the `date()` method and is currently not supported when calling `DateTime::createFromFormat()`. This provides a (hacky) work-around, so I'd appreciate any feedback on putting together a better solution.

I was unable to get this solely working using `date()` which failed to work where timezones were used in the date format specifically for characters `e`, `T`, `P` and in turn would fail on the respective unit tests. See further details on my last comment on the issue https://github.com/laravel/framework/issues/19445#issuecomment-306491065

I'm a newb to contributing so would appreciate it if you kept that in mind when providing feedback. Thank you!

### CODE REVIEW

File: src/Illuminate/Validation/Concerns/ValidatesAttributes.php [hunk 1]
```diff
@@ -358,7 +358,12 @@ protected function validateDateFormat($attribute, $value, $parameters)
             return false;
         }
 
-        $date = DateTime::createFromFormat($parameters[0], $value);
+        if (!preg_match('/v/', $parameters[0])) {
```
Comment (tillkruss) [inline src/Illuminate/Validation/Concerns/ValidatesAttributes.php]: `preg_match()` isn't necessary here. Just use `strpos()` or `Str::contains()`.

File: src/Illuminate/Validation/Concerns/ValidatesAttributes.php [hunk 2]
```diff
@@ -358,7 +358,12 @@ protected function validateDateFormat($attribute, $value, $parameters)
...
+            $date = DateTime::createFromFormat($parameters[0], $value);
+        } else {
+            $dateString = date($parameters[0], strtotime($value));
+            $date = new DateTime($dateString);
+        }
```
Comment (tillkruss) [inline src/Illuminate/Validation/Concerns/ValidatesAttributes.php]: `$dateString` is unnecessary and can be inlined.

Comment (tillkruss) [inline src/Illuminate/Validation/Concerns/ValidatesAttributes.php]: I'd even write it like this:

```php
$date = strpos($parameters[0], 'v') === false
    ? DateTime::createFromFormat($parameters[0], $value)
    : new DateTime(date($parameters[0], strtotime($value)))
```

Comment (haakym) [inline src/Illuminate/Validation/Concerns/ValidatesAttributes.php]: @tillkruss Thanks a lot for your feedback - most appreciated! I've updated the code to reflect your feedback.

**Comment by taylorotwell** [comment]
Why is it not supported when using DateTime::createFromFormat? Is PHP fixing this?

**Comment by haakym** [comment]
@taylorotwell Searching the web hasn't provided me with any answers as to why or if a fix is in the works. Any recommendations on where to further investigate? Is this something that should be raised on https://bugs.php.net?

**Comment by morloderex** [comment]
@taylorotwell I belive this has been fixed in php 7.1. :)

https://github.com/php/php-src/blob/PHP-7.1/NEWS#L645

So maybe we should add a check for that?

**Comment by haakym** [comment]
@morloderex I don't see anything in the link about `milliseconds` only `createFromFormat()` and `microseconds`, am I missing something?

**Comment by morloderex** [comment]
@haakym 

For instance this is fixed in php 7.1 
Fixed bug #52514 (microseconds are missing in DateTime class). (Derick)

**Comment by taylorotwell** [comment]
This feels more like a PHP bug. Going to close for the time being and if you need it you can create a custom rule extension.

**Comment by Kyslik** [comment]
I personally don't even know whats practical use of validating milliseconds (or doing any kind of operation with millisecond precision).

Anyone care to share?

**Pull request closed without merging**